File: 496-next-greater-element-i/next-greater-element-i.js
var nextGreaterElement = function(nums1, nums2) {
    let lookup = {}
    let stack = []
    let result = new Array(nums1.length).fill(-1)

    for (let i = 0; i < nums1.length; i++) {
        let char = nums1[i]
        lookup[char] = i
    }    

    for (let i = nums2.length - 1; i >= 0; i--) {
        let char = nums2[i]

        if (!stack.length) {
            stack.push(char)
        }

        while (stack.length && char >= stack.at(-1)) {
            stack.pop()
        }
        if (stack.length) {
            if (char in lookup) {
                index = lookup[char]
                result[index] = stack.at(-1)
            }

        }
        stack.push(char)
      
    }
    return result
};

File: 496-next-greater-element-i/next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        result = [-1] * len(nums1)
        lookup = {}
        stack = []

        for idx, val in enumerate(nums1):
            lookup[val] = idx

        j = len(nums2) - 1

        while (j >= 0):
            if (not stack):
                stack.append(nums2[j])
            else:
                while (stack and nums2[j] >= stack[-1]):
                    stack.pop()
                if (stack and nums2[j] in lookup):
                    idx = lookup[nums2[j]]
                    result[idx] = stack[-1]
                stack.append(nums2[j])

            j -= 1

        return result

        